File: year-2/programming-design/java/MotorcycleContestClientServer/MotorcycleClient/src/main/java/services/MotorcycleServiceProxy.java
package services;

import domain.Participant;
import domain.Race;
import domain.User;
import dtos.CredentialsDto;
import dtos.ParticipantDto;
import dtos.RaceDto;
import dtos.UserDto;
import networking.Request;
import networking.RequestType;
import networking.Response;
import networking.ResponseType;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

public class MotorcycleServiceProxy extends BaseServiceProxy implements MotorcycleService {
    private static final Logger LOGGER = LogManager.getLogger(BaseServiceProxy.class);
    private MotorcycleObserver observer;

    public MotorcycleServiceProxy(String host, int port) {
        super(host, port);
    }

    public MotorcycleServiceProxy() {
    }

    @Override
    public Optional<User> authenticate(String username, String password) throws Exception {
        LOGGER.traceEntry("authenticating account by name " + username + " and password " + password);
        this.getConnection();
        UserDto userDto = new UserDto(username, password);
        Request request = new Request.Builder().type(RequestType.LOGIN).data(userDto).build();
        LOGGER.info("sending login request {}", request);
        this.sendRequest(request);
        Response response = readResponse();
        LOGGER.info("got login response {}", response);
        if (response.type() == ResponseType.OK){
            User user = (User) response.data();
            if (user == null) {
                LOGGER.info("user not found");
                return Optional.empty();
            }
            LOGGER.info("found user {}", user);
            return Optional.of(user);
        }
        if (response.type() == ResponseType.ERROR){
            String exception = response.data().toString();
            closeConnection();
            this.observer = null;
            LOGGER.traceExit("exception getting races");
            throw new ServerException(exception);
        }
        return Optional.empty();
    }

    @Override
    public RaceDto getRaceByCapacity(Integer capacity) throws Exception {
        LOGGER.traceEntry("entering get race by capacity {}", capacity);
        this.getConnection();
        Request request = new Request.Builder().type(RequestType.FILTER_RACES).data(capacity).build();
        LOGGER.info("sending get race by capacity {} request", capacity);
        sendRequest(request);
        Response response = readResponse();
        LOGGER.info("get race by capacity response {}", response);
        if (response.type() == ResponseType.OK){
            RaceDto race = (RaceDto) response.data();
            LOGGER.info("get race by capacity found {}", race);
            return race;
        }
        if (response.type() == ResponseType.ERROR){
            String exception = response.data().toString();
            LOGGER.traceExit("exception getting race by capacity {}", capacity);
            throw new ServerException(exception);
        }
        return null;
    }

    @Override
    public List<RaceDto> getRaceInfos() throws Exception {
        LOGGER.traceEntry("entering get races information");
        this.getConnection();
        Request request = new Request.Builder().type(RequestType.SEARCH_RACES).build();
        LOGGER.info("sending get races information request {}", request);
        sendRequest(request);
        Response response = readResponse();
        LOGGER.info("get races information response {}", response);
        if (response.type() == ResponseType.OK){
            List<RaceDto> races = (ArrayList<RaceDto>) response.data();
            LOGGER.info("get races information found {}", races);
            return races;
        }
        if (response.type() == ResponseType.ERROR){
            String exception = response.data().toString();
            LOGGER.traceExit("exception getting races information");
            throw new ServerException(exception);
        }
        return null;
    }

    @Override
    public Iterable<Race> getRaces() throws Exception {
        LOGGER.traceEntry("entering get all races");
        this.getConnection();
        Request request = new Request.Builder().type(RequestType.GET_RACES).build();
        LOGGER.info("sending get races request {}", request);
        sendRequest(request);
        Response response = readResponse();
        LOGGER.info("get races response {}", response);
        if (response.type() == ResponseType.OK){
            List<Race> races = (ArrayList<Race>) response.data();
            LOGGER.info("get races found {}", races);
            return races;
        }
        if (response.type() == ResponseType.ERROR){
            String exception = response.data().toString();
            LOGGER.traceExit("exception finding user");
            throw new ServerException(exception);
        }
        return null;
    }

    @Override
    public List<ParticipantDto> getRaceParticipantsByTeam(String teamName) throws Exception {
        LOGGER.traceEntry("entering get participants by team {}", teamName);
        this.getConnection();
        Request request = new Request.Builder().type(RequestType.GET_PARTICIPANTS).data(teamName).build();
        LOGGER.info("sending get participants by team {} request", teamName);
        sendRequest(request);
        Response response = readResponse();
        LOGGER.info("get get participants by team response {}", response);
        if (response.type() == ResponseType.OK){
            List<ParticipantDto> participants = (List<ParticipantDto>) response.data();
            LOGGER.info("get get participants by team found {}", participants);
            return participants;
        }
        if (response.type() == ResponseType.ERROR){
            String exception = response.data().toString();
            LOGGER.traceExit("exception getting participants by team {}", teamName);
            throw new ServerException(exception);
        }
        return null;
    }

    @Override
    public void registerParticipant(String name, String teamName, Integer capacity) throws Exception {
        LOGGER.traceEntry("entering register participant {}, {}, {}", name, teamName, capacity);
        this.getConnection();
        Request request = new Request.Builder()
                .type(RequestType.REGISTER)
                .data(new CredentialsDto(name, teamName, capacity))
                .build();
        LOGGER.info("sending register participant request");
        sendRequest(request);
        Response response = readResponse();
        LOGGER.info("register participant response {}", response);
        if (response.type() == ResponseType.OK){
            LOGGER.info("register participant successful");
        }
        else if (response.type() == ResponseType.ERROR){
            String exception = response.data().toString();
            LOGGER.traceExit("exception during register participant");
            throw new ServerException(exception);
        }
    }

    @Override
    public void logOut(User user) {
        if (connection == null){
            LOGGER.info("connection already closed");
            return;
        }
        LOGGER.traceEntry("logging out account with name " + user.getName());
        UserDto userDto = new UserDto(user.getUsername(), user.getPassword());
        Request request = new Request.Builder().type(RequestType.LOGOUT).data(userDto).build();
        LOGGER.info("sending logout request {}", request);
        sendRequest(request);
        Response response = readResponse();
        closeConnection();
        this.observer = null;
        if (response.type() == ResponseType.ERROR){
            String exception = response.data().toString();
            closeConnection();
            LOGGER.traceExit("exception closing connection");
            throw new ServerException(exception);
        }
    }

    protected void handleRegisteredParticipant(Response response) {
        Participant participant = (Participant) response.data();
        observer.registeredParticipant(participant);
    }

    @Override
    public void addMotorcycleObserver(MotorcycleObserver observer) {
        this.observer = observer;
    }

    @Override
    public void removeMotorcycleObserver(MotorcycleObserver observer) {
        this.observer = null;
    }
}

File: year2/mpp/c#/MotorcycleClientServer/MotorcycleClient/Services/MotorcycleServiceProxy.cs
using System;
using System.Collections.Generic;
using MotorcycleCommon.Networking;
using MotorcycleCommon.Services;
using MotorcycleContest.Domain.Dtos;
using MotorcycleContest.Domain.Entities;

namespace motorcycle_contest.Services
{
    public class MotorcycleServiceProxy : BaseServiceProxy, IMotorcycleService
    {
        private IMotorcycleObserver _observer;
        
        public MotorcycleServiceProxy(string host, int port) : base(host, port)
        {
        }
        
        public User Authenticate(string username, string password)
        {
            Logger.InfoFormat("autheticating account by name {0} and password {1}", username, password);
            GetConnection();
            Request request = new Request()
            {
                Type = RequestType.LOGIN,
                Data = new UserDto(username, password)
            };
            Logger.InfoFormat("sending login request {0}", request);
            SendRequest(request);
            Response response = ReadResponse();
            Logger.InfoFormat("got login response {0}", response);
            switch (response.Type)
            {
                case ResponseType.OK:
                {
                    User user = (User) response.Data;
                    if (user == null)
                    {
                        Logger.Info("user not found");
                        return null;
                    }
                    Logger.InfoFormat("found user {0}", user);
                    return user;
                }
                case ResponseType.ERROR:
                {
                    String exception = response.Data.ToString();
                    CloseConnection();
                    _observer = null;
                    Logger.Info("exception getting races");
                    throw new ServerException(exception);
                }
                case ResponseType.REGISTERED_PARTICIPANTS:
                    break;
                default:
                    return null;
            }
            return null;
        }

        public RaceDto GetRaceByCapacity(int capacity)
        {
            Logger.InfoFormat("entering get race by capacity {0}", capacity);
            GetConnection();
            Request request = new Request()
            {
                Type = RequestType.FILTER_RACES,
                Data = capacity
            };
            Logger.InfoFormat("sending get race by capacity request {0}", request);
            SendRequest(request);
            Response response = ReadResponse();
            Logger.InfoFormat("get race by capacity response {0}", response);
            switch (response.Type)
            {
                case ResponseType.OK:
                {
                    RaceDto race = (RaceDto) response.Data;
                    Logger.InfoFormat("get race by capacity found {0}", race);
                    return race;
                }
                case ResponseType.ERROR:
                {
                    String exception = response.Data.ToString();
                    Logger.Info("exception get race by capacity");
                    throw new ServerException(exception);
                }
                default:
                    return null;
            }
        }

        public IEnumerable<Race> GetRaces()
        {
            Logger.Info("entering get races");
            GetConnection();
            Request request = new Request()
            {
                Type = RequestType.GET_RACES
            };
            Logger.InfoFormat("sending get racesrequest {0}", request);
            SendRequest(request);
            Response response = ReadResponse();
            Logger.InfoFormat("get races response {0}", response);
            switch (response.Type)
            {
                case ResponseType.OK:
                {
                    List<Race> races = (List<Race>) response.Data;
                    Logger.InfoFormat("get races found {0}", races);
                    return races;
                }
                case ResponseType.ERROR:
                {
                    String exception = response.Data.ToString();
                    Logger.Info("exception get races");
                    throw new ServerException(exception);
                }
                default:
                    return null;
            }
        }

        public List<ParticipantDto> GetRaceParticipantsByTeam(string teamName)
        {
            Logger.InfoFormat("entering get race participants by team {0}", teamName);
            GetConnection();
            Request request = new Request()
            {
                Type = RequestType.GET_PARTICIPANTS,
                Data = teamName
            };
            Logger.InfoFormat("sending get race participants by team {0}", request);
            SendRequest(request);
            Response response = ReadResponse();
            Logger.InfoFormat("get race participants by team response {0}", response);
            switch (response.Type)
            {
                case ResponseType.OK:
                {
                    List<ParticipantDto> participants = (List<ParticipantDto>) response.Data;
                    Logger.InfoFormat("get race participants by team found {0}", participants);
                    return participants;
                }
                case ResponseType.ERROR:
                {
                    String exception = response.Data.ToString();
                    Logger.Info("exception get race participants by team");
                    throw new ServerException(exception);
                }
                default:
                    return null;
            }
        }

        public void RegisterParticipant(string name, string teamName, int capacity)
        {
            Logger.InfoFormat("entering register participant with name {0}, teamName {1}, capacity {2}", name, teamName, capacity);
            GetConnection();
            Request request = new Request()
            {
                Type = RequestType.REGISTER,
                Data = new CredentialsDto(name, teamName, capacity)
            };
            Logger.InfoFormat("sending register participant request {0}", request);
            SendRequest(request);
            Response response = ReadResponse();
            Logger.InfoFormat("register participant response {0}", response);
            switch (response.Type)
            {
                case ResponseType.OK:
                {
                    Logger.Info("register participant successful");
                    break;
                }
                case ResponseType.ERROR:
                {
                    String exception = response.Data.ToString();
                    Logger.Info("exception during register participant");
                    throw new ServerException(exception);
                }
            }
        }
        
        public override void HandleRegisteredParticipant(Response response)
        {
            Logger.InfoFormat("handling registered participant response {0}", response);
            Participant participant = (Participant) response.Data;
            _observer.RegisteredParticipant(participant);
        }

        public void logOut(User user)
        {
            Logger.Info("logging out");
            if (_connection == null)
            {
                Logger.Info("connection already closed");
                return;
            }

            Request request = new Request()
            {
                Type = RequestType.LOGOUT
            };
            SendRequest(request);
            Response response = ReadResponse();
            CloseConnection();
            if (response.Type == ResponseType.ERROR)
            {
                String exception = response.Data.ToString();
                Logger.Warn("logging out failed " + exception);
                throw new ServerException(exception);
            }
        }

        public void AddMotorcycleObserver(IMotorcycleObserver observer)
        {
            this._observer = observer;
        }

        public void RemoveMotorcycleObserver(IMotorcycleObserver observer)
        {
            this._observer = null;
        }
    }
}
